### PR TITLE
Fix dtype inference during gradient norm computation

### PR DESCRIPTION
**Pull request by carmocca**
## What does this PR do?

Fixes https://github.com/Lightning-AI/lightning/issues/14020

### Does your PR introduce any breaking changes? If yes, please list them.

None

## Before submitting

- [x] Was this **discussed/approved** via a GitHub issue? (not for typos and docs)
- [x] Did you read the [contributor guideline](https://github.com/Lightning-AI/lightning/blob/master/.github/CONTRIBUTING.md), **Pull Request** section?
- [x] Did you make sure your **PR does only one thing**, instead of bundling different changes together?
- [n/a] Did you make sure to **update the documentation** with your changes? (if necessary)
- [x] Did you write any **new necessary tests**? (not for typos and docs)
- [x] Did you verify new and **existing tests pass** locally with your changes?
- [x] Did you list all the **breaking changes** introduced by this pull request?
- [x] Did you **update the CHANGELOG**? (not for typos, docs, test updates, or minor internal changes/refactors)

## PR review

- [x] Is this pull request ready for review? (if not, please submit in draft mode)
- [x] Check that all items from **Before submitting** are resolved
- [x] Make sure the title is self-explanatory and the description concisely explains the PR
- [x] Add labels and milestones (and optionally projects) to the PR so it can be classified

cc @borda @akihironitta

### CODE REVIEW

**Review by rohitgr7 — APPROVED** [review]

**Review by awaelchli — APPROVED** [review]

**Comment by fshart** [comment]
The proposed changes won't fix the mentioned bug, as the line
`total_norm = torch.tensor(list(norms.values())).norm(norm_type)`
(line 49 in src/pytorch_lightning/utilities/grads.py) causes the issue because the
`torch.tensor`
forces an intermediate typecast (to torch.float32).

**Comment by fshart** [comment]
As mentioned in https://github.com/Lightning-AI/lightning/issues/14020, there is a type interference needed for 'max_dtype', i.e. a type to which all norms can be typecast without becoming infinite.

**Comment by carmocca** [comment]
@fshart The issue is that `list(norms.values())` was a list of primitive floats after calling `.item()` on their respective tensors, so when creating the "total" tensor, PyTorch would typecast to float32. However, with this patch, we defer the `.item()` call to the end so the `dtype` is correctly kept as `float64` for the total norm computation.

If I'm missing something, could you come up with a failing test using the changes in this branch?

**Comment by fshart** [comment]
@carmocca I stand corrected. The MWE from the original issue works with bugfix.

**Review by akihironitta — APPROVED** [review]
LGTM!

**Review by Borda — APPROVED** [review]
